From 2ea75a7e0801231eaccb112a8935beaffd26f956 Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Thu, 28 Aug 2014 14:22:08 -0400
Subject: [cpufreq] Fix wrong time unit conversion

Message-id: <1409235879-1283-35-git-send-email-prarit@redhat.com>
Patchwork-id: 89495
O-Subject: [RHEL7.1 PATCH BZ 1134639 034/185] cpufreq: Fix wrong time unit conversion
Bugzilla: 1134369
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Mark Langsdorf <mlangsdo@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1134639

commit a857c0b9e24e39fe5be82451b65377795f9538d8
Author: Andreas Schwab <schwab@linux-m68k.org>
Date:   Sat Sep 7 18:35:08 2013 +0200

    cpufreq: Fix wrong time unit conversion

    The time spent by a CPU under a given frequency is stored in jiffies unit
    in the cpu var cpufreq_stats_table->time_in_state[i], i being the index of
    the frequency.

    This is what is displayed in the following file on the right column:

         cat /sys/devices/system/cpu/cpuX/cpufreq/stats/time_in_state
         2301000 19835820
         2300000 3172
         [...]

    Now cpufreq converts this jiffies unit delta to clock_t before returning it
    to the user as in the above file. And that conversion is achieved using the API
    cputime64_to_clock_t().

    Although it accidentally works on traditional tick based cputime accounting, where
    cputime_t maps directly to jiffies, it doesn't work with other types of cputime
    accounting such as CONFIG_VIRT_CPU_ACCOUNTING_* where cputime_t can map to nsecs
    or any granularity preffered by the architecture.

    For example we get a buggy zero delta on full dyntick configurations:

         cat /sys/devices/system/cpu/cpuX/cpufreq/stats/time_in_state
         2301000 0
         2300000 0
         [...]

    Fix this with using the proper jiffies_64_t to clock_t conversion.

    Reported-and-tested-by: Carsten Emde <C.Emde@osadl.org>
    Signed-off-by: Andreas Schwab <schwab@linux-m68k.org>
    Signed-off-by: Frederic Weisbecker <fweisbec@gmail.com>
    Acked-by: Paul E. McKenney <paulmck@linux.vnet.ibm.com>
    Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>

Cc: Lenny Szubowicz <lszubowi@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/cpufreq/cpufreq_stats.c b/drivers/cpufreq/cpufreq_stats.c
index 4d7eb36..2eaa066 100644
--- a/drivers/cpufreq/cpufreq_stats.c
+++ b/drivers/cpufreq/cpufreq_stats.c
@@ -75,7 +75,7 @@ static ssize_t show_time_in_state(struct cpufreq_policy *policy, char *buf)
  for (i = 0; i < stat->state_num; i++) {
   len += sprintf(buf + len, "%u %llu\n", stat->freq_table[i],
    (unsigned long long)
-   cputime64_to_clock_t(stat->time_in_state[i]));
+   jiffies_64_to_clock_t(stat->time_in_state[i]));
  }
  return len;
 }
-- 
1.7.1